### PR TITLE
provide custom implementation of metal::partition (for performance)

### PR DESCRIPTION
**Pull request by ecrypa**

### CODE REVIEW

**Comment by brunocodutra** [comment]
@ecrypa I notice this is still a draft, what's missing?

File: include/metal/list/partition.hpp [hunk 1]
```diff
@@ -36,8 +41,73 @@ namespace metal {
...
+            using type = list<list<>, list<val>>;
+        };
+
+        template<class conds, class seq>
+        struct _partitioner {};
+
+        template<int_... ns, class... vals>
+        struct _partitioner<list<number<ns>...>, list<vals...>>
+            : _partition_joiner<
+                  typename _partition_filter<ns>::template type<vals>...> {};
```
Comment (brunocodutra) [inline include/metal/list/partition.hpp]: Have we tested `if_<number<ns>, pair<list<val>, list<>>, pair<list<>, list<val>>>...` vs `_partition_filter`?

Comment (ecrypa) [reply include/metal/list/partition.hpp]: > Have we tested `if_<number<ns>, pair<list<val>, list<>>, pair<list<>, list<val>>>...` vs `_partition_filter`?

I am not absolutely sure about that. I will benchmark it.

I remember why I once introduced that `filter`: What you propose here is similar to [this pattern](https://github.com/brunocodutra/metal/pull/115/commits/34eeae4d64dfe52e015b169b82ce166be2b9f0bb#diff-2a7fabd857009f87aacd803768e57228R59), which is the ["GCC9 killer (v0)"](https://ecrypa.github.io/metal-benchmark/nth_element_gcc9/). I thus introduced the indirection via [`_fast_partition_helper`](https://github.com/brunocodutra/metal/pull/115/commits/34eeae4d64dfe52e015b169b82ce166be2b9f0bb#diff-d77f2ed31417a4569b562d1540f6dbcbR21) in v2, which is faster on GCC9. When extracting the partitioner, I probably renamed `_fast_partition_helper` into `filter`.

Comment (brunocodutra) [reply include/metal/list/partition.hpp]: I'm happy to merge this as is either way and iterate on it later if necessary, unless you have plans for this PR still?

Comment (ecrypa) [reply include/metal/list/partition.hpp]: Good suggestion! Feel free to merge.

**Pull request merged**